### PR TITLE
Astparser (DO-NOT-MERGE)

### PR DESCRIPTION
**Pull request by ingvagabund**
Just notes

### CODE REVIEW

File: pkg/types/symboltable/symboltable.go [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+type SymbolDef struct {
+	Pos  DeclPos          `json:"-"`
+	Name string           `json:"name"`
+	Def  gotypes.DataType `json:"def,omitempty"`
+}
+
+
+func (sym *SymbolDef) GetPosition() DeclPos { return sym.Pos }
+func (sym *SymbolDef) GetName() string      { return sym.Name }
+func (sym *SymbolDef) GetType() string      { return sym.Def.GetType() }
```
Comment (ingvagabund) [inline pkg/types/symboltable/symboltable.go:33]: Those functions are not necessary as long as there is nothing like `Symbol` interface and the symbol is basically an attribute. We can access all fields directly as long as they are exported. E.g.

```go
s := &SymbolDef{
   Name: "Name",
   Pos: DeclPos{
      File: "file.go",
      Line: 55,
   },
   Def: nil,
}

fmt.Printf("Name: %v\n", s.Name)
fmt.Printf("Pos: %v\n", s.Pos)
fmt.Printf("Def: %#v\n", s.Def)
```

File: pkg/types/symboltable/symboltable.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+//TODO: maybe something like this would be part of types.go? Used just
+//      to find data type of processed symbol
+//TODO: signature... would be fine use same as json.Unmarshal
+func unmarshalDataType(b []byte) (gotypes.DataType, error) {
```
Comment (ingvagabund) [inline pkg/types/symboltable/symboltable.go:38]: This is already generated https://github.com/gofed/symbols-extractor/commit/6223b32617476e16b5dcb402af5513236cee0fc5

File: pkg/types/symboltable/symboltable.go [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+		}
+		symbol = &t
+	default:
+		return nil, fmt.Errorf("JSON unmarshal: Unknown symbol type '%v'", symbolType)
+	}
+
+	return symbol, nil
+}
+
+func (o *SymbolDef) UnmarshalJSON(b []byte) error {
```
Comment (ingvagabund) [inline pkg/types/symboltable/symboltable.go:144]: this is already generated https://github.com/gofed/symbols-extractor/commit/6223b32617476e16b5dcb402af5513236cee0fc5

File: pkg/types/symboltable/symboltable.go [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+	} else if sym, ok := o.datatypes[id]; ok {
+		return sym
+	} else if sym, ok := o.constants[id]; ok {
+		return sym
+	}
+
+	return nil
+}
+
+func (o *Package) Init() {
```
Comment (ingvagabund) [inline pkg/types/symboltable/symboltable.go:198]: Use ``func NewPackage() *Package {...}`` instead.

File: pkg/types/symboltable/symboltable.go [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+}
+
+///////////////////////////////////////////////////////////////////
+
+type Package struct {
+	//TODO: really?
+	variables map[string]*SymbolDef
+	functions map[string]*SymbolDef
+	datatypes map[string]*SymbolDef
+	constants map[string]*SymbolDef
```
Comment (ingvagabund) [inline pkg/types/symboltable/symboltable.go:172]: There is no difference between variable and constant from a PoV of a symbol definition.

File: pkg/types/symboltable/symboltable.go [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+	if _, ko := o.functions[id]; ko {
+		return fmt.Errorf(errSymExists, sym.Name, sym)
+	}
+
+	o.functions[id] = sym
+	o.Functions = append(o.Functions, sym)
+	return nil
+}
+
+func (o *Package) AddVariable(id string, sym *SymbolDef) error {
```
Comment (ingvagabund) [inline pkg/types/symboltable/symboltable.go:226]: The body of the function is almost the same as of the ``AddFunction``.  You could have:

```go
const ( 
        VariableSymbol = iota
        FunctionSymbol = iota
        DataTypeSymbol = iota
)

type Package struct {
   symbols map[int]map[string]*gotypes.SymbolDef
   Symbols map[int][]*gotypes.SymbolDef
}

func NewPackage() *Package {
   return &Package{
      symbols: map[int]map[string]*gotypes.SymbolDef{
         VariableSymbol: make(map[string]*gotypes.SymbolDef, 0),
         FunctionSymbol: make(map[string]*gotypes.SymbolDef, 0),
         DataTypeSymbol: make(map[string]*gotypes.SymbolDef, 0),
      },
      Symbols: map[int][]*gotypes.SymbolDef{
         VariableSymbol: make([]*gotypes.SymbolDef, 0),
         FunctionSymbol: make([]*gotypes.SymbolDef, 0),
         DataTypeSymbol: make([]*gotypes.SymbolDef, 0),
      },
   }
}

func (o *Package) AddSymbol(symbolType int, id string, sym *SymbolDef) error {
        if _, ko := o.symbols[symbolType][id]; ko {
		return fmt.Errorf(errSymExists, sym.Name, sym)
	}

	o.symbols[symbolType][id] = sym
	o.Symbols[symbolType] = append(o.Symbols[symbolType], sym)
	return nil
}

func (o *Package) AddVariable(id string, sym *SymbolDef) error {
        return o.AddSymbol(VariableSymbol, id, sym)
}

func (o *Package) AddDataType(id string, sym *SymbolDef) error {
        return o.AddSymbol(DataTypeSymbol, id, sym)
}

...
```

**Comment by pirat89** [comment]
obsoleted

**Pull request closed without merging**